packages/cli/src/webhooks/test-webhooks.ts

```diff
@@ -218,25 +219,46 @@ export class TestWebhooks implements IWebhookManager {
 	 * Return whether activating a workflow requires listening for webhook calls.
 	 * For every webhook call to listen for, also activate the webhook.
 	 */
-	async needsWebhook(
-		userId: string,
-		workflowEntity: IWorkflowDb,
-		additionalData: IWorkflowExecuteAdditionalData,
-		runData?: IRunData,
-		pushRef?: string,
-		destinationNode?: string,
-	) {
+	async needsWebhook(options: {
+		userId: string;
+		workflowEntity: IWorkflowDb;
+		additionalData: IWorkflowExecuteAdditionalData;
+		runData?: IRunData;
+		pushRef?: string;
+		destinationNode?: string;
+		preferredTrigger?: WorkflowRequest.ManualRunPayload['preferredTrigger'];
+	}) {
+		const userId = options.userId;
+		const workflowEntity = options.workflowEntity;
+		const additionalData = options.additionalData;
+		const runData = options.runData;
+		const pushRef = options.pushRef;
+		const destinationNode = options.destinationNode;
+		const preferredTrigger = options.preferredTrigger;
+
 		if (!workflowEntity.id) throw new WorkflowMissingIdError(workflowEntity);
 
 		const workflow = this.toWorkflow(workflowEntity);
 
-		const webhooks = WebhookHelpers.getWorkflowWebhooks(
+		let webhooks = WebhookHelpers.getWorkflowWebhooks(
 			workflow,
 			additionalData,
 			destinationNode,
 			true,
 		);
 
+		// If we have a preferred trigger with data, we don't have to listen for a
+		// webhook.
+		if (preferredTrigger?.data) {
+			return false;
+		}
+
+		// If we have a preferred trigger without data we only want to listen for
+		// that trigger, not the other ones.
+		if (preferredTrigger) {
+			webhooks = webhooks.filter((w) => w.node === preferredTrigger.name);
+		}
```

Comment:
I don't follow. 

This check is removing all but one webhook from the list if the preferred trigger was passed. 

The next check is checking if the the list contains at least one trigger that is restartable (I think 🤔 )

Reply:
I guess the question is, can it ever happen that a trigger to start triggers the restartable check? What exactly is `w.webhookDescription.restartWebhook`?

But if not worth it let's leave it :)

Reply:
> What exactly is w.webhookDescription.restartWebhook?

I don't know either tbh. Jan added this here: https://github.com/n8n-io/n8n/commit/5a179cd5aebb57d9f5017bb71544370af4953327#diff-b386248ff00977749c873ed85821c241b773e9740d7e7adf94e05b73b350ed74

Which is a huge commit. 

It seems this is for the Form, Wait, Slack and Gmail nodes.

> I guess the question is, can it ever happen that a trigger to start triggers the restartable check?

I guess that would be fine still. I'm assuming this is to not listen for webhook for workflows that have been put to into waiting? 

Which means you can't actually pick them as a trigger to start from.

At least for the the wait and form nodes this is only interesting for workflows that wait, probably for slack too, e.g. if you render a dialog and wait for the user to click a button. I'm not sure why the gmail node offers this though.

I guess this can be left until the split button is implemented.

cc @cstuncsik @CharlieKolb